### PR TITLE
docgen: move to shared RST state (fix #16990)

### LINKED ISSUE #16990: rst: runnableExamples affects rst underline headings
### Example
```nim
## Basic Usage
## ===========
##
## some comment
##
## Lists1
## ------
##
## Lists2
## ------
##
runnableExamples:
  discard

## Lists3
## ------
##
runnableExamples:
  discard


## Lists4
## ------
runnableExamples:
  discard


## # Lists5
runnableExamples:
  discard


## ## Lists6
runnableExamples:
  discard

##[

# Lists7
## List 8
## List 9
foo
]##
```

### Current Output
image

### Expected Output
* `Lists4 ------` is a bug, should appear as `List2`
* `Lists3` is a bug, should appear as `List2`

everything else works fine

### workaround
use markdown style prefix form (list5,...,list9) instead of underline form as was done in https://github.com/nim-lang/Nim/pull/16981
(not really a workaround since prefix form is IMO better style, more familiar for markdown users, and easier to use since you don't have to match the length of title with underline markers)

### Additional Information
1.5.1 0d34345f29aa3b862f21c8dcff6b94b5ac5b9572
/cc @a-mr since you've fixed many rst issues recently :)

### PR DESCRIPTION
**Pull request by a-mr**
Previously each doc comment inside a `.nim` file was processed as a separate document.

This PR:

A. makes doc comments related through a common (per-file) RST shared state. (Strictly speaking this would be enough for fixing #16990)
B. splits RST processing into 2 phases:
  1. pass 1 where `string` doc comment are converted to RST AST `PRstNode`
  2. pass 2 (`finishGenerateDoc`), which traverses RST AST and performs final resolution (headings + links + footnotes + RST substitutions) converting `PRstNode` to HTML or Latex `string`

B. is done by saving `PRstNode` of any doc comment to a data structure after pass 1.

So point B. makes `.nim` doc comments feel like a single document. This style becomes possible:

```nim
## Ref lnk_.
##
## See lnk_ and [1]_ for |nimlang|.

proc f*() =
  ## See again lnk_ and [1]_ for |nimlang|.
  discard

## .. _lnk: http://example.org/
## .. [1] Footnote 1.
## .. |nimlang| replace:: Nim programming language
```

This PR aids in implementing smart links nim-lang/RFCs#125, because previously doc comments were converted immediately, that is before the whole `.nim` AST tree were traversed and hence the list of all local symbols (`proc`s etc) were not yet known. Now `finishGenerateDoc` will be the right place to resolve all the links.

cc @timotheecour

### CODE REVIEW

**Comment by timotheecour** [comment]
just a side-note, @a-mr, please write `fix #16990` in the PR body (in title is useful but not enough, and github doesn't recognize `fixing #16990`) so that github cross-references it properly (eg so that it'll show "may be fixed by ...` in that issue); as noted in contributing.html

File: compiler/main.nim [hunk 1]
```diff
@@ -281,7 +281,7 @@ proc mainCommand*(graph: ModuleGraph) =
     else:
       rawMessage(conf, errGenerated, "'run' command not available; rebuild with -d:tinyc")
   of cmdDoc0: docLikeCmd commandDoc(cache, conf)
-  of cmdDoc2:
+  of cmdDoc2:  # the 'nim doc' command (aka doc2)
```
Comment (timotheecour) [inline compiler/main.nim]: or simply rename cmdDoc2 to cmdDoc, making it self-documenting; note also that such comments are best next to symbol definition instead of symbol use

File: lib/packages/docutils/rst.nim [hunk 2]
```diff
@@ -508,19 +508,19 @@ type
                                   # number, order of occurrence
     msgHandler: MsgHandler      # How to handle errors.
     findFile: FindFileHandler   # How to find files.
+    filename: string
+    hasToc*: bool
 
-  PSharedState = ref SharedState
+  RstSharedState* = ref SharedState
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim]: don't violate the convention `type PFoo = ref Foo` though, if you do that, then also rename SharedState to RstSharedState, ditto with newSharedState => newRstSharedState (not newSharedStateRst)

(if you want, you can make a separate PR with just those renames maybe that can be reviewed fast; looks like those symbols are all private anyways)

File: lib/packages/docutils/rst.nim [hunk 3]
```diff
@@ -578,35 +578,41 @@ proc whichRoleAux(sym: string): RstNodeKind =
...
     result = rnUnknownRole
 
-proc newSharedState(options: RstParseOptions,
-                    findFile: FindFileHandler,
-                    msgHandler: MsgHandler): PSharedState =
+proc newSharedStateRst*(options: RstParseOptions,
+                        filename: string,
+                        findFile: FindFileHandler,
+                        msgHandler: MsgHandler): RstSharedState =
   new(result)
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim]: in this or (preferably i guess) separate cleanup PR:
result = RstSharedState(a1: a1, a2: a2, ...)

and these are not needed anymore (relic from the past):
`result.subs = @[]`

File: lib/packages/docutils/rst.nim [hunk 4]
```diff
@@ -578,35 +578,41 @@ proc whichRoleAux(sym: string): RstNodeKind =
...
   if not fileExists(result):
     result = p.s.findFile(filename)
 
 proc rstMessage(p: RstParser, msgKind: MsgKind, arg: string) =
-  p.s.msgHandler(p.filename, curLine(p),
+  p.s.msgHandler(p.s.filename, curLine(p),
                              p.col + currentTok(p).col, msgKind, arg)
 
+proc rstMessage(s: RstSharedState, msgKind: MsgKind, arg: string) =
+  ## Handle message `arg`. Its position can not be localized at the moment.
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim]: doc comment unclear, can you try to explain `can not be localized` better in the doc comment?

File: lib/packages/docutils/rst.nim [hunk 5]
```diff
@@ -754,14 +758,14 @@ proc rstnodeToRefname(n: PRstNode): string =
   var b = false
   rstnodeToRefnameAux(n, result, b)
 
-proc findSub(p: var RstParser, n: PRstNode): int =
+proc findSub(s: RstSharedState, n: PRstNode): int =
   var key = addNodes(n)
   # the spec says: if no exact match, try one without case distinction:
-  for i in countup(0, high(p.s.subs)):
-    if key == p.s.subs[i].key:
+  for i in countup(0, high(s.subs)):
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim:765]: it's ok to cleanup lines that you are editing; `0..high(s.subs)` is shorter and better

File: lib/packages/docutils/rst.nim [hunk 6]
```diff
@@ -838,28 +842,28 @@ proc addFootnoteSymAuto(p: var RstParser) =
   p.s.lineFootnoteSym.add curLine(p)
   p.s.footnotes.add((fnAutoSymbol, -1, -1, p.s.lineFootnoteSym.len, ""))
 
-proc orderFootnotes(p: var RstParser) =
+proc orderFootnotes(s: var RstSharedState) =
   ## numerate auto-numbered footnotes taking into account that all
   ## manually numbered ones always have preference.
-  ## Save the result back to p.s.footnotes.
+  ## Save the result back to s.footnotes.
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim]:
```suggestion
  ## Save the result back to `s.footnotes`.
```

File: tests/stdlib/trstgen.nim [hunk 7]
```diff
@@ -1034,7 +1034,8 @@ Test1
       """
     var warnings8 = new seq[string]
     let output8 = input8.toHtml(warnings=warnings8)
-    check(warnings8[] == @["input(4, 1) Warning: unknown substitution " &
+    # TODO: the line 1 is arbitrary because reference lines are not preserved
```
Comment (timotheecour) [inline tests/stdlib/trstgen.nim:1037]: is that a regression or was `4` meaningless/wrong? (i know that line numbers were often wrong in nim rst2html)

File: compiler/docgen.nim [hunk 8]
```diff
@@ -28,24 +28,50 @@ const
   DocColOffset = "## ".len  # assuming that a space was added after ##
 
 type
+  ItemFragment = object  ## A fragment from each item will be eventually
+                         ## constructed by converting `rst` fields to strings.
+    case isRst: bool
+    of true:
+      rst: PRstNode
+    of false:
```
Comment (timotheecour) [inline compiler/docgen.nim]: add a comment indicating what it corresponds to if isRst is false

File: compiler/docgen.nim [hunk 9]
```diff
@@ -28,24 +28,50 @@ const
...
+      str: string
+  ItemPre = seq[ItemFragment]  ## A pre-processed item.
+  Item = object        ## Any item in documentation, e.g. type or
+                       ## proc entry. Configuration variable ``doc.item``
```
Comment (timotheecour) [inline compiler/docgen.nim]:
```suggestion
  Item = object        ## Any item in documentation, e.g. symbol
                       ## entry. Configuration variable ``doc.item``
```

?
(because of const, let etc)

File: compiler/docgen.nim [hunk 10]
```diff
@@ -28,24 +28,50 @@ const
...
+                       ## is used for its HTML rendering.
+    descRst: ItemPre     ## Description of the item (may contain runnable
```
Comment (timotheecour) [inline compiler/docgen.nim]: call a cat a cat (`runnableExamples`), more searchable

File: compiler/docgen.nim [hunk 11]
```diff
@@ -28,24 +28,50 @@ const
...
+                         ## examples).
+    substitutions: seq[string]    ## Variable names in ``doc.item``...
```
Comment (timotheecour) [inline compiler/docgen.nim]:
```suggestion
    substitutions: seq[string]    ## Variable names in `doc.item`...
```

File: compiler/docgen.nim [hunk 12]
```diff
@@ -840,9 +868,9 @@ proc genItem(d: PDoc, n, nameNode: PNode, k: TSymKind, docFlags: DocFlags) =
 
   let seeSrc = genSeeSrc(d, toFullPath(d.conf, n.info), n.info.line.int)
 
-  d.section[k].add(getConfigVar(d.conf, "doc.item") %
-    ["name", name, "uniqueName", uniqueName,
-     "header", result, "desc", comm, "itemID", $d.id,
```
Comment (timotheecour) [inline compiler/docgen.nim:845]: why was `"desc", comm` removed? (no need for a doc comment but just for review)

File: compiler/docgen.nim [hunk 13]
```diff
@@ -886,48 +914,49 @@ proc genItem(d: PDoc, n, nameNode: PNode, k: TSymKind, docFlags: DocFlags) =
...
+proc genJsonItem(d: PDoc, n, nameNode: PNode, k: TSymKind): JsonItem =
   if not isVisible(d, nameNode): return
   var
     name = getName(d, nameNode)
     comm = genRecComment(d, n)
     r: TSrcGen
   initTokRender(r, n, {renderNoBody, renderNoComments, renderDocComments})
-  result = %{ "name": %name, "type": %($k), "line": %n.info.line.int,
-                 "col": %n.info.col}
+  result.json = %{ "name": %name, "type": %($k), "line": %n.info.line.int,
```
Comment (timotheecour) [inline compiler/docgen.nim:924]: in future work, IMO we should use jsonutils here, doing a similar simplification as was done in https://github.com/nim-lang/Nim/pull/18100, which allows more type-safety (via a type that represents the spec of what's serialized, making deserialization + serialization easier)

File: compiler/docgen.nim [hunk 14]
```diff
@@ -1034,13 +1063,16 @@ proc documentRaises*(cache: IdentCache; n: PNode) =
     if p5 != nil: n[pragmasPos].add p5
 
 proc generateDoc*(d: PDoc, n, orig: PNode, docFlags: DocFlags = kDefault) =
+  ## Goes through nim nodes recursively and collects doc comments.
+  ## Main function for `doc`:option: (aka `doc2`:option:) command,
```
Comment (timotheecour) [inline compiler/docgen.nim]:
```suggestion
  ## Main function for `doc`:option: command,
```

(doc2 is now an undocumented alias to doc and all references should disappear; and even then, it shouldn't be re-defined here)

File: compiler/docgen.nim [hunk 15]
```diff
@@ -1357,6 +1434,7 @@ proc commandRst2TeX*(cache: IdentCache, conf: ConfigRef) =
   commandRstAux(cache, conf, conf.projectFull, TexExt)
 
 proc commandJson*(cache: IdentCache, conf: ConfigRef) =
+  ## implementation of a deprecated jsondoc0 command
```
Comment (timotheecour) [inline compiler/docgen.nim:1435]: then should we rename commandJson as commandJson0 ? (can be done in future PR)

File: compiler/docgen.nim [hunk 16]
```diff
@@ -1079,21 +1111,65 @@ proc generateDoc*(d: PDoc, n, orig: PNode, docFlags: DocFlags = kDefault) =
...
+  preparePass2(d.sharedState, firstRst)
+
+  # Finalize fragments of ``.nim`` or ``.rst`` file
+  proc renderItemPre(d: PDoc, fragments: ItemPre, result: var string) =
+    for f in fragments:
+      case f.isRst:
+      of true:
+        var resolved = resolveSubs(d.sharedState, f.rst)
+        var str: string
+        renderRstToOut(d[], resolved, str)
```
Comment (timotheecour) [inline compiler/docgen.nim]: can we make this work? (ie, make sure that renderRstToOut doesn't clear its 3rd arg but insteads just appeands, as is good practice)
`renderRstToOut(d[], resolved, result)`

**Comment by timotheecour** [comment]
* @a-mr I'm inclined to accept it as-is, I tested it locally and went through everything and the comments I have can be addressed in a followup cleanup PR (to avoid making this PR bitrot for too long)

* #16990 now does work but a test case can be added in followup PR

* it wasn't mentioned in PR description, but IIUC this also helps with `nim jsondoc` ? is that being tested?

**Review by timotheecour — APPROVED** [review]
LGTM, remaining points can be addressed in followup PR

File: tests/stdlib/trstgen.nim (hunk 7 above)
Comment (a-mr) [reply tests/stdlib/trstgen.nim:1037]: not a regression: it's wrong in both cases. Before it was always last line of text, after it's always first line. The problem is that line/column info has never been preserved in any references :-(. We need to add it to `(P)RstNode`, a good task for one of future PRs.

File: compiler/docgen.nim (hunk 15 above)
Comment (a-mr) [reply compiler/docgen.nim:1435]: I thought about deleting this `jsondoc0` altogether. It always crashes for me, including stable 1.4 version. Does it work for you?

**Comment by a-mr** [comment]
> this also helps with nim jsondoc?

I tested `jsondoc` manually, yes, it works as expected.

As i wrote above `jsondoc0` always crashes for me (incl. 1.4).

**Comment by timotheecour** [comment]
does it also fix any of these ?
https://github.com/nim-lang/Nim/issues?q=is%3Aissue+is%3Aopen+jsondoc

File: compiler/docgen.nim (hunk 15 above)
Comment (timotheecour) [reply compiler/docgen.nim:1435]: crashes with `options.nim(637, 3) `not conf.outFile.isEmpty`  [AssertionDefect]` but that doesn't mean it'd be hard to fix; if this were to be removed, it should be a separate PR though (IIRC the use case is for include files where semcheck don't work; just like nim doc0; it's hacky but ./koch docs relies on at least nim doc0)

File: lib/packages/docutils/rst.nim [hunk 17]
```diff
@@ -838,28 +842,28 @@ proc addFootnoteSymAuto(p: var RstParser) =
   p.s.lineFootnoteSym.add curLine(p)
   p.s.footnotes.add((fnAutoSymbol, -1, -1, p.s.lineFootnoteSym.len, ""))
 
-proc orderFootnotes(p: var RstParser) =
+proc orderFootnotes(s: var RstSharedState) =
```
Comment (Araq) [inline lib/packages/docutils/rst.nim]:
```suggestion
proc orderFootnotes(s: RstSharedState) =
```

It's a ref, it's already mutable and you don't reassign the pointer, do you?

File: lib/packages/docutils/rst.nim [hunk 18]
```diff
@@ -1765,17 +1769,18 @@ proc getLevel(p: var RstParser, c: char, hasOverline: bool): int =
                             line: curLine(p), hasPeers: false)
   result = p.s.hLevels.len - 1
 
-proc countTitles(p: var RstParser, n: PRstNode) =
-  ## Fill `p.s.hTitleCnt`
+proc countTitles(s: var RstSharedState, n: PRstNode) =
```
Comment (Araq) [inline lib/packages/docutils/rst.nim]:
```suggestion
proc countTitles(s: RstSharedState, n: PRstNode) =
```

File: lib/packages/docutils/rst.nim [hunk 19]
```diff
@@ -2806,40 +2809,60 @@ proc parseDotDot(p: var RstParser): PRstNode =
...
+  ## The result should be further processed by
+  ## `preparePass2` and `resolveSubs` (which is pass 2).
+  var p: RstParser
+  initParser(p, sharedState)
+  p.line = line
+  p.col = column
+  getTokens(fragment, p.tok)
+  result = parseDoc(p)
+
+proc preparePass2*(s: var RstSharedState, mainNode: PRstNode) =
```
Comment (Araq) [inline lib/packages/docutils/rst.nim]:
```suggestion
proc preparePass2*(s: RstSharedState, mainNode: PRstNode) =
```

File: lib/packages/docutils/rst.nim (hunk 5 above)
Comment (a-mr) [reply lib/packages/docutils/rst.nim:765]: agree but let us to do it in follow-up

File: compiler/main.nim (hunk 1 above)
Comment (a-mr) [reply compiler/main.nim]: agree, done.

File: lib/packages/docutils/rst.nim (hunk 2 above)
Comment (a-mr) [reply lib/packages/docutils/rst.nim]: ok, ok, i don't argue. Done.

File: lib/packages/docutils/rst.nim (hunk 3 above)
Comment (a-mr) [reply lib/packages/docutils/rst.nim]: ok, done.

File: lib/packages/docutils/rst.nim (hunk 4 above)
Comment (a-mr) [reply lib/packages/docutils/rst.nim]: rephrased it.

File: compiler/docgen.nim (hunk 8 above)
Comment (a-mr) [reply compiler/docgen.nim]: added.

File: compiler/docgen.nim (hunk 10 above)
Comment (a-mr) [reply compiler/docgen.nim]: ok.

File: compiler/docgen.nim (hunk 12 above)
Comment (a-mr) [reply compiler/docgen.nim:845]: doc comment `comm` changed type (was `string` before, became `PRstNode`) and cannot be added together with strings. It became a separate field `Item.descRst`.

File: compiler/docgen.nim (hunk 15 above)
Comment (a-mr) [reply compiler/docgen.nim:1435]: may be but i'm not planning on doing any fixes, because i don't understand the use cases of `jsondoc/jsondoc0` commands in the 1st place :-/

File: compiler/docgen.nim (hunk 16 above)
Comment (a-mr) [reply compiler/docgen.nim]: yes, well spotted, thanks.

File: lib/packages/docutils/rst.nim (hunk 2 above)
Comment (Clyybber) [reply lib/packages/docutils/rst.nim]: > don't violate the convention type PFoo = ref Foo though

That's an old and outdated convention.

File: compiler/docgen.nim (hunk 15 above)
Comment (timotheecour) [reply compiler/docgen.nim:1435]: `nim jsondoc` use case is for tooling, so that a program can use the json output generated by nim for further processing (eg to support queries, such as find all procs matching a given signature) without having to link against nim or use compiler API, refs https://github.com/nim-lang/Nim/issues?q=is%3Aissue+is%3Aopen+jsondoc.

But yes, defering to future work is good
